### DISCUSSION TITLE
익명함수와 람다 표현식에 대해서 설명해주세요

### DISCUSSION THREAD

**Discussion by kunsanglee** [2024-woowacourse-study/level-interview#61] [Java]
.

**Comment by gyungchan-jo** [comment]
익명 함수(Anonymous Function):

이름이 없는 함수로, 보통 함수의 인자로 전달되거나 반환값으로 사용됩니다.
익명 함수는 함수형 언어에서 자주 사용됩니다.
Java에서는 익명 내부 클래스를 사용하여 익명 함수를 흉내내는 것이 가능합니다.

람다 표현식(Lambda Expression):

익명 함수의 간결한 문법적인 변형으로, Java 8부터 도입되었습니다.
함수형 인터페이스를 구현하는 코드 블록을 간단하게 표현할 수 있습니다.
함수형 프로그래밍에서 많이 사용되며, 코드의 가독성을 향상시킵니다.

따라서 익명 함수와 람다 표현식은 기본적으로 비슷한 개념이지만, 람다 표현식은 간결한 문법을 제공하여 코드를 더욱 간단하게 만들어줍니다. 특히 Java에서는 람다 표현식을 이용하여 함수형 프로그래밍의 장점을 활용할 수 있습니다.

**Comment by le2sky** [comment]
**(익명 클래스와 람다식에 대해서 설명합니다.)**

익명 클래스는 이름이 없는 클래스입니다. 
선언한 부분에서만 인스턴스를 생성할 수 있으며, 클래스의 이름이 필요한 작업은 수행할 수 없습니다.
인터페이스를 구현하는 동시에 다른 클래스를 상속받을 수 없으며, 여러 인터페이스를 구현할 수 없습니다.
또한, 익명 클래스가 길어지는 경우에는 가독성이 떨어질 수 있습니다. 
자바에서는 함수 타입을 표현할 때, 추상 메서드를 하나만 담은 인터페이스를 사용했습니다. 이 인터페이스를 구현하는
인스턴스를 함수 객체라고 합니다. 람다가 생기기 이전에는 주로 작은 함수 객체를 만드는 데 익명 클래스를 자주 사용되었습니다.

 Java 8부터 추상 메서드 하나를 가진 인터페이스를 함수형 인터페이스라고 부릅니다. 해당 인터페이스의 인스턴스를 람다식을 사용해 만들 수 있도록 지원되었습니다. 람다식의 간결한 문법을 이용하여, 가독성 좋은 코드를 작성할 수 있습니다.
하지만, 람다를 이용할 때는 몇가지 주의해야 합니다. 다음은 람다식 사용 시 유의해야할 부분입니다.

람다식은 메서드나 클래스와 달리 이름이 없고 문서화가 어렵습니다. 따라서, 람다식 자체로 동작이 명확히 설명되어야 합니다.
함수형 인터페이스가 아닌 곳에서는 사용할 수 없습니다. 이럴 경우에는 익명 클래스를 사용해야 합니다.
람다는 자신을 참조할 수 없습니다. 람다 내부에서 this 키워드는 외부 인스턴스를 가리킵니다.

**Comment by alstn113** [comment]
## 익명 함수
익명 함수를 구현하기 위해서 익명 내부 클래스를 사용한다.
인터페이스를 구현하거나 추상 클래스를 확장하는 익명 클래스를 생성하는 것이다.
```java
Runnable runnable = new Runnable() {
    @Override
    public void run() {
        System.out.println("This is an anonymous function.");
    }
};
```

## 람다 표현식
람다 표현식은 함수형 인터페이스의 인스턴스를 생성하는 간결한 방법을 제공한다.
함수형 인터페이스는 하나의 추상 메서드만을 가지는 인터페이스이다.
람다 표현식은 이러한 함수형 인터페이스를 구현하기 위한 익명 메서드를 정의한다.
```java
Runnable runnable = () -> {
    System.out.println("This is a lambda expression.");
};
```

**Comment by GIVEN53** [comment]
익명 내부 클래스 사용
인터페이스 구현, 추상 클래스 확장

함수형 인터페이스 인스턴스 생성
